src/eredis_cluster_manager.erl

```diff
@@ -0,0 +1,415 @@
+%% @doc Get key for command where the key is in th 4th position (eval and
+%% evalsha commands)
+%% @end
+%% =============================================================================
+-spec get_key_from_rest([anystring()]) -> string() | undefined.
+get_key_from_rest([_,KeyName|_]) when is_bitstring(KeyName) ->
+    bitstring_to_list(KeyName);
+get_key_from_rest([_,KeyName|_]) when is_list(KeyName) ->
+    KeyName;
+get_key_from_rest(_) ->
+    undefined.
+
+%% =============================================================================
+%% @doc Return the hash slot from the key
+%% @end
+%% =============================================================================
+-spec get_key_slot(Key::any()) -> Slot::integer().
+get_key_slot(K) ->
+    %% cast any type of key to a list.
+    Key = lists:concat([K]),
+    KeyToBeHashed = case string:chr(Key,${) of
+                        0 ->
+                            Key;
+                        Start ->
+                            case string:chr(string:substr(Key,Start+1),$}) of
+                                0 ->
+                                    Key;
+                                Length ->
+                                    if
+                                        Length =:= 1 ->
+                                            Key;
+                                        true ->
+                                            string:substr(Key,Start+1,Length-1)
+                                    end
+                            end
+                    end,
+    eredis_cluster_hash:hash(KeyToBeHashed).
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Methods to manage list of eredis connections
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+init_eredis_connections() ->
+    ets:new(?MODULE, [protected, set, named_table, {read_concurrency, true}]).
+
+%% Looks up existing redis client pid using Ip, Port and return Pid of the client
+%% connection. If no such connection is present, this method establishes the connection.
+-spec lookup_eredis_pid(Ip::string(), Port::integer()) -> {ok, Pid::pid()}.
+lookup_eredis_pid(Ip, Port) ->
+    case ets:lookup(?MODULE, [Ip, Port]) of
+        [] ->
+           {ok, Pid} = safe_eredis_start_link(Ip, Port),
+           ets:insert(?MODULE, {[Ip, Port], Pid}),
+          {ok, Pid};
+        [_, Pid] -> {ok, Pid}
+    end.
+
+
+%% Closes redis client connection if present in cstate.
+%%-spec delete(Ip::string(), Port::integer()) -> {ok}.
+%%delete(Ip, Port) ->
+%%    case ets:lookup(?MODULE, [Ip, Port]) of
+%%        [_, Pid] ->
+%%             safe_eredis_stop(Pid),
+%%             ets:remove(?MODULE, [Ip, Port]),
+%%            ok; 
+%%        {_, _} -> {ok}
+%%    end. 
+
+
+safe_eredis_start_link(Ip, Port) ->
+    process_flag(trap_exit, true),
+    Payload = eredis:start_link(Ip, Port),
+    process_flag(trap_exit, false),
+    Payload.
```

Comment:
Can you explain what this does?

Reply:
Here's my understanding:
Executing process_flag in process1 basically traps exits signals from other process that are linked to it. This results in our process1 not crashing when the linked processes exit: since the exit signal is transformed into another message. If we do not trap exit signals: then when the linked process crashes: our process will crash as well.

Reply:
@balusu-murali explanation is accurate.

I don't understand it fully yet. I used the current code in eredis_cluster as a guideline to write this code. I will spend a bit more time to understand and explain in a day or so.

Reply:
They basically do not want this cluster_manager gen_server process to not die if there was any issue with the eredis client process. They probably aim to discover that the eredis client process is not working properly later and attempt to reconnect it.

I think these exit messages need to be handled somewhere?

Reply:
What we want is our gen_server never to die if one of the eredis connections crashes. We need to trap signals from all the eredis process, or alternatively we should somehow create them as monitor where eredis clients die if we die, but if they die we don't die.

Reply:
yes, exactly! that is what I was suggesting. So we need to trap these exit messages from the process that die. I think there probably is some code to do this already in eredis_cluster.erl ?

Reply:
Added some code to handle these errors. Not perfect. Let me know what you think.